pallets/server/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+
+			Self::deposit_event(Event::Registered(who, server));
+			Ok(().into())
+		}
+
+		#[pallet::weight(T::WeightInfo::transfer_owner())]
+		pub fn transfer_owner(
+			origin: OriginFor<T>,
+			server_id: HashOf<T>,
+			new_owner: AccountIdOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			match <Self as ServerInterface<T>>::transfer_owner(&server_id, &who, &new_owner) {
+				Ok(server) => {
+					Self::deposit_event(Event::OwnerTransferred(who, server.owner, server.id));
+					Ok(().into())
+				},
+				Err(error) => Err(error.into()),
+			}
+		}
+
+		#[pallet::weight(T::WeightInfo::update_name(new_name.len() as u32))]
+		pub fn update_name(
+			origin: OriginFor<T>,
+			server_id: HashOf<T>,
+			new_name: Vec<u8>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			match <Self as ServerInterface<T>>::update_name(&server_id, &who, &new_name) {
+				Ok(server) => {
+					Self::deposit_event(Event::NameUpdated(who, server.name, server.id));
+					Ok(().into())
+				},
+				Err(error) => Err(error.into()),
+			}
+		}
+
+		#[pallet::weight(T::WeightInfo::unregister())]
+		pub fn unregister(
+			origin: OriginFor<T>,
+			server_id: ServerIdOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			match <Self as ServerInterface<T>>::unregister(&server_id, &who) {
+				Ok(_) => {
+					Self::deposit_event(Event::Unregistered(who, server_id));
+					Ok(().into())
+				},
+				Err(error) => Err(error.into()),
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn generate_server_id(owner: &T::AccountId, name: &[u8]) -> HashOf<T> {
+			let mut seed = owner.encode();
+			let account_info = frame_system::Pallet::<T>::account(owner);
+
+			seed.append(&mut account_info.nonce.encode());
+			seed.append(&mut name.encode());
+
+			T::Hashing::hash(&seed)
+		}
+	}
+
+	impl<T: Config> ServerInterface<T> for Pallet<T> {
+		type Error = Error<T>;
+		type Server = ServerOf<T>;
+		type Name = Vec<u8>;
+
+		fn get_by_id(server_id: &T::Hash) -> Option<Self::Server> {
+			Self::server_by_id(server_id)
+		}
+
+		fn register(account_id: &T::AccountId, name: &Self::Name) -> Self::Server {
+			let id = Self::generate_server_id(account_id, name);
+			let server = Server::new(id, account_id.clone(), name.to_vec());
+
+			ServerById::<T>::insert(id, server.clone());
+
+			server
+		}
+
+		fn transfer_owner(
+			server_id: &T::Hash,
+			account_id: &T::AccountId,
+			new_owner: &T::AccountId,
+		) -> Result<Self::Server, Self::Error> {
+			if !ServerById::<T>::contains_key(server_id) {
+				return Err(Error::<T>::NotExists)
+			}
+
+			let mut server = <Self as ServerInterface<T>>::get_by_id(server_id).unwrap();
+			let current_owner = server.get_owner().clone();
+
+			if &current_owner != account_id {
+				return Err(Error::<T>::Unauthorized)
+			}
+
+			if &current_owner == new_owner {
+				return Err(Error::<T>::OwnerNotChanged)
```

Comment:
this should success response, return current server, this only avoid unnecessary storage update